src/sql-parser/src/lexer.rs

```diff
@@ -0,0 +1,358 @@
+// directly into Materialize on December 21, 2019.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License in the LICENSE file at the
+// root of this repository, or online at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! SQL lexer.
+//!
+//! This module lexes SQL according to the rules described in the ["Lexical
+//! Structure"] section of the PostgreSQL documentation. The description is
+//! intentionally not replicated here. Please refer to that chapter as you
+//! read the code in this module.
+//!
+//! ["Lexical Structure"]: https://www.postgresql.org/docs/current/sql-syntax-lexical.html
+
+use std::convert::TryFrom;
+use std::ops::Range;
+
+use crate::keywords::ALL_KEYWORDS;
+use crate::lexer::buf::LexBuf;
+use crate::parser::ParserError;
+
+mod buf;
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum Token {
+    Word(Word),
+    String(String),
+    HexString(String),
+    Number(String),
+    Parameter(usize),
+    Op(String),
+    Star,
+    Eq,
+    LParen,
+    RParen,
+    LBracket,
+    RBracket,
+    Dot,
+    Comma,
+    Colon,
+    DoubleColon,
+    Semicolon,
+}
+
+impl Token {
+    pub fn name(&self) -> &str {
+        match self {
+            Token::Word(w) => w.name(),
+            Token::String(_) => "string literal",
+            Token::HexString(_) => "hex string literal",
+            Token::Number(_) => "number",
+            Token::Parameter(_) => "parameter",
+            Token::Op(_) => "operator",
+            Token::Star => "star",
+            Token::Eq => "equals sign",
+            Token::LParen => "left parenthesis",
+            Token::RParen => "right parenthesis",
+            Token::LBracket => "left square bracket",
+            Token::RBracket => "right square bracket",
+            Token::Dot => "dot",
+            Token::Comma => "comma",
+            Token::Colon => "colon",
+            Token::DoubleColon => "double colon",
+            Token::Semicolon => "semicolon",
+        }
+    }
+}
+
+/// A keyword (like SELECT) or an optionally quoted SQL identifier.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Word {
+    /// The value of the word, without the enclosing quotes, and with the
+    /// escape sequences (if any) processed.
+    pub value: String,
+    /// Whether the word was quoted.
+    pub quoted: bool,
+    /// If the word was not quoted and it matched one of the known keywords,
+    /// this will have one of the values from dialect::keywords, otherwise empty
+    pub keyword: String,
```

Comment:
Yep, the design of this is really silly! It should really be an enum, so that we can match on an int instead of a string. But I want to leave that to another commit, because this one is big enough as it is, and changing this struct will involve a lot of parser changes.